File: frontend/src/app/services/worker.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Worker, WorkerRequest } from '../models/worker.models';

@Injectable({
  providedIn: 'root'
})
export class WorkerService {
  private readonly API_URL = 'http://localhost:30000/api/workers';

  constructor(private http: HttpClient) {}

  getAllWorkers(): Observable<Worker[]> {
    return this.http.get<Worker[]>(this.API_URL);
  }

  getWorkerById(id: number): Observable<Worker> {
    return this.http.get<Worker>(`${this.API_URL}/${id}`);
  }

  createWorker(request: WorkerRequest): Observable<Worker> {
    return this.http.post<Worker>(this.API_URL, request);
  }

  updateWorker(id: number, request: WorkerRequest): Observable<Worker> {
    return this.http.put<Worker>(`${this.API_URL}/${id}`, request);
  }

  deleteWorker(id: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${id}`);
  }

  searchWorkers(name: string): Observable<Worker[]> {
    return this.http.get<Worker[]>(`${this.API_URL}/search?name=${name}`);
  }

  getWorkersByPosition(position: string): Observable<Worker[]> {
    return this.http.get<Worker[]>(`${this.API_URL}/by-position?position=${position}`);
  }
}

File: backend/src/main/java/com/equipepay/service/WorkerService.java
package com.equipepay.service;

import com.equipepay.dto.WorkerRequest;
import com.equipepay.dto.WorkerResponse;
import com.equipepay.entity.Worker;
import com.equipepay.repository.WorkerRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class WorkerService {

    private final WorkerRepository workerRepository;

    public WorkerService(WorkerRepository workerRepository) {
        this.workerRepository = workerRepository;
    }

    @Transactional
    public WorkerResponse createWorker(WorkerRequest request) {
        if (workerRepository.existsByPhone(request.getPhone())) {
            throw new RuntimeException("Phone number already exists");
        }

        Worker worker = new Worker();
        worker.setName(request.getName());
        worker.setPhone(request.getPhone());
        worker.setPosition(request.getPosition());
        worker.setMonthlySalary(request.getMonthlySalary());
        worker.setJoinDate(request.getJoinDate());
        worker.setPhotoUrl(request.getPhotoUrl());

        Worker savedWorker = workerRepository.save(worker);
        return convertToResponse(savedWorker);
    }

    @Transactional
    public WorkerResponse updateWorker(Long id, WorkerRequest request) {
        Worker worker = workerRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Worker not found"));

        if (!worker.getPhone().equals(request.getPhone()) && workerRepository.existsByPhone(request.getPhone())) {
            throw new RuntimeException("Phone number already exists");
        }

        worker.setName(request.getName());
        worker.setPhone(request.getPhone());
        worker.setPosition(request.getPosition());
        worker.setMonthlySalary(request.getMonthlySalary());
        worker.setJoinDate(request.getJoinDate());
        worker.setPhotoUrl(request.getPhotoUrl());

        Worker updatedWorker = workerRepository.save(worker);
        return convertToResponse(updatedWorker);
    }

    @Transactional
    public void deleteWorker(Long id) {
        if (!workerRepository.existsById(id)) {
            throw new RuntimeException("Worker not found");
        }
        workerRepository.deleteById(id);
    }

    @Transactional(readOnly = true)
    public WorkerResponse getWorkerById(Long id) {
        Worker worker = workerRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Worker not found"));
        return convertToResponse(worker);
    }

    @Transactional(readOnly = true)
    public List<WorkerResponse> getAllWorkers() {
        return workerRepository.findAll().stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<WorkerResponse> searchWorkersByName(String name) {
        return workerRepository.findByNameContaining(name).stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<WorkerResponse> getWorkersByPosition(String position) {
        return workerRepository.findByPosition(position).stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }

    private WorkerResponse convertToResponse(Worker worker) {
        return new WorkerResponse(
                worker.getId(),
                worker.getName(),
                worker.getPhone(),
                worker.getPosition(),
                worker.getMonthlySalary(),
                worker.getDailySalary(),
                worker.getJoinDate(),
                worker.getPhotoUrl(),
                worker.getCreatedAt(),
                worker.getUpdatedAt()
        );
    }
}
